scripts/dats_jsonld_annotator/annotator.py

```diff
@@ -0,0 +1,323 @@
+import argparse
+import json
+import logging
+import pathlib as pal
+import sys
+import time
+
+import jsonschema as jss
+
+
+CONTEXT_DIR = (pal.Path(__file__).parent / "context" / "sdo").resolve()
+SCHEMA_DIR = (pal.Path(__file__).parent / "schema").resolve()
+logger = logging.getLogger("DATS annotator")
+
+
+def find_schema(parent_schema, term, json_object):
+    """
+    This function finds the appropriate JSON schema for a term in a DATS Json object. To do
+    so, it relies on the JSON schema of the supplied JSON object.
+
+    :param parent_schema: The DATS JSON schema that corresponds to the json_object parameter
+    :param term: the string term for which the schema shall be found
+    :param json_object:  the JSON object that contains the term and to which the parent_schema corresponds
+    :return: the DATS JSON schema for the term
+    """
+    ps = parent_schema["properties"]
+    if term not in ps.keys():
+        # TODO: Decide whether to raise an Exception here
+        # The only way this could happen on a valid dataset is if someone has specified additional properties
+        # We'd like to discourage this
+        logger.debug(
+            f'I cannot find {term = } in the parent schema: {parent_schema["id"]}.\n{json_object = }'
+        )
+        return None
+
+    # Set up the schema resolver
+    resolver = jss.RefResolver(base_uri=parent_schema["id"], referrer=None)
+
+    # See if "items" is in the schema entry
+    search_dict = ps[term]
+    if "items" in search_dict.keys():
+        search_dict = ps[term]["items"]
+
+    if "$ref" in search_dict.keys():
+        schema_name = search_dict["$ref"]
+    elif len(set(search_dict.keys()).intersection(["anyOf", "oneOf", "allOf"])) > 0:
+        # Get the correct jsonld relationship
+        schema_rel = list(
+            set(search_dict.keys()).intersection(["anyOf", "oneOf", "allOf"])
+        )[0]
+        possible_keys = []
+        for ref in search_dict[schema_rel]:
+            # Get the schema
+            ref_name = ref["$ref"]
+            _schema_uri, _schema = resolver.resolve(ref_name)
+            if _schema_uri == resolver.base_uri:
+                _schema = parent_schema
+            if jss.Draft4Validator(_schema).is_valid(json_object):
+                possible_keys.append(ref_name)
+        if len(possible_keys) > 1:
+            # TODO: decide if we let the user pick which option to go with
+            logger.debug(f"I got more than one option for {term}: {possible_keys}")
+        elif len(possible_keys) == 0:
```

Comment:
I think this could get indeed very verbose. There will be thousands of "matches". We could add it and log at the debug level if you think this would be an important information to keep around